File: Packages/StrandAnalytics/Sources/StrandAnalytics/HeartRateRecovery.swift
import Foundation
import WhoopProtocol

/// Heart-rate recovery after a sufficiently intense workout (#516).
///
/// The engine deliberately works from the recorded HR stream rather than the workout's editable summary
/// fields. It looks for sustained Zone-3-or-higher effort near the end of the session, takes the highest
/// recorded HR in the final 30 seconds as the cessation value, then compares it with robust median readings
/// around 1, 2 and 5 minutes after the workout ended.
///
/// Missing post-workout coverage stays nil. The engine never interpolates across a gap or turns a missing
/// reading into zero, which is important when a strap disconnects as the workout ends.
public enum HeartRateRecovery {
    public struct Result: Equatable, Sendable {
        public let endHR: Int
        public let after1Minute: Int?
        public let after2Minutes: Int?
        public let after5Minutes: Int?

        public init(endHR: Int, after1Minute: Int?, after2Minutes: Int?, after5Minutes: Int?) {
            self.endHR = endHR
            self.after1Minute = after1Minute
            self.after2Minutes = after2Minutes
            self.after5Minutes = after5Minutes
        }

        public var hasMeasurement: Bool {
            after1Minute != nil || after2Minutes != nil || after5Minutes != nil
        }
    }

    /// Zone 3 starts at 70% HRmax in NOOP's display-zone model.
    public static let eligibilityFractionOfMaxHR = 0.70
    /// The high-intensity effort must be sustained, not a single optical spike.
    public static let minimumHighIntensitySeconds = 120
    /// Eligibility is intentionally tied to the end of the workout: HRR is meaningful only when exercise
    /// cessation, rather than a long cool-down, anchors the recovery clock.
    public static let eligibilityLookbackSeconds = 300
    public static let cessationWindowSeconds = 30
    public static let measurementToleranceSeconds = 15
    public static let minimumSamplesPerReading = 3
    public static let maximumContinuousGapSeconds = 10

    public static func calculate(samples: [HRSample], workoutStart: Int, workoutEnd: Int,
                                 maxHR: Double) -> Result? {
        guard workoutStart > 0, workoutEnd > workoutStart, maxHR > 0 else { return nil }
        let lowerBound = max(workoutStart, workoutEnd - eligibilityLookbackSeconds)
        let upperBound = workoutEnd + 5 * 60 + measurementToleranceSeconds
        let sorted = samples
            .filter { $0.ts >= lowerBound && $0.ts <= upperBound && (30...250).contains($0.bpm) }
            .sorted { lhs, rhs in lhs.ts == rhs.ts ? lhs.bpm < rhs.bpm : lhs.ts < rhs.ts }
        guard sorted.count >= minimumSamplesPerReading else { return nil }

        let beforeEnd = sorted.filter { $0.ts <= workoutEnd }
        let threshold = maxHR * eligibilityFractionOfMaxHR
        guard sustainedSeconds(atOrAbove: threshold, in: beforeEnd) >= minimumHighIntensitySeconds else {
            return nil
        }

        let cessation = beforeEnd
            .filter { $0.ts >= workoutEnd - cessationWindowSeconds }
            .map(\.bpm)
        guard cessation.count >= minimumSamplesPerReading, let endHR = cessation.max() else { return nil }

        func recovery(at minutes: Int) -> Int? {
            let target = workoutEnd + minutes * 60
            let values = sorted
                .filter { abs($0.ts - target) <= measurementToleranceSeconds }
                .map(\.bpm)
            guard values.count >= minimumSamplesPerReading, let reading = median(values) else { return nil }
            return endHR - reading
        }

        let result = Result(
            endHR: endHR,
            after1Minute: recovery(at: 1),
            after2Minutes: recovery(at: 2),
            after5Minutes: recovery(at: 5)
        )
        return result.hasMeasurement ? result : nil
    }

    private static func sustainedSeconds(atOrAbove threshold: Double, in samples: [HRSample]) -> Int {
        guard samples.count >= 2 else { return 0 }
        var seconds = 0
        for i in 0..<(samples.count - 1) {
            let gap = samples[i + 1].ts - samples[i].ts
            guard gap > 0, gap <= maximumContinuousGapSeconds else { continue }
            if Double(samples[i].bpm) >= threshold { seconds += gap }
        }
        return seconds
    }

    private static func median(_ values: [Int]) -> Int? {
        guard !values.isEmpty else { return nil }
        let sorted = values.sorted()
        let middle = sorted.count / 2
        if sorted.count.isMultiple(of: 2) {
            return Int((Double(sorted[middle - 1] + sorted[middle]) / 2.0).rounded())
        }
        return sorted[middle]
    }
}

File: android/app/src/main/java/com/noop/analytics/HeartRateRecovery.kt
package com.noop.analytics

import com.noop.data.HrSample
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Heart-rate recovery after a sufficiently intense workout (#516). Kotlin twin of
 * `StrandAnalytics/HeartRateRecovery.swift`.
 *
 * Missing post-workout coverage stays null: the engine never interpolates across a disconnect or turns
 * a missing reading into zero.
 */
object HeartRateRecovery {
    data class Result(
        val endHr: Int,
        val after1Minute: Int?,
        val after2Minutes: Int?,
        val after5Minutes: Int?,
    ) {
        val hasMeasurement: Boolean
            get() = after1Minute != null || after2Minutes != null || after5Minutes != null
    }

    const val eligibilityFractionOfMaxHr = 0.70
    const val minimumHighIntensitySeconds = 120L
    const val eligibilityLookbackSeconds = 300L
    const val cessationWindowSeconds = 30L
    const val measurementToleranceSeconds = 15L
    const val minimumSamplesPerReading = 3
    const val maximumContinuousGapSeconds = 10L

    fun calculate(samples: List<HrSample>, workoutStart: Long, workoutEnd: Long, maxHr: Double): Result? {
        if (workoutStart <= 0L || workoutEnd <= workoutStart || maxHr <= 0.0) return null
        val lowerBound = maxOf(workoutStart, workoutEnd - eligibilityLookbackSeconds)
        val upperBound = workoutEnd + 5 * 60 + measurementToleranceSeconds
        val sorted = samples
            .filter { it.ts in lowerBound..upperBound && it.bpm in 30..250 }
            .sortedWith(compareBy<HrSample> { it.ts }.thenBy { it.bpm })
        if (sorted.size < minimumSamplesPerReading) return null

        val beforeEnd = sorted.filter { it.ts <= workoutEnd }
        val threshold = maxHr * eligibilityFractionOfMaxHr
        if (sustainedSeconds(threshold, beforeEnd) < minimumHighIntensitySeconds) return null

        val cessation = beforeEnd
            .filter { it.ts >= workoutEnd - cessationWindowSeconds }
            .map { it.bpm }
        if (cessation.size < minimumSamplesPerReading) return null
        val endHr = cessation.maxOrNull() ?: return null

        fun recovery(minutes: Int): Int? {
            val target = workoutEnd + minutes * 60L
            val values = sorted.filter { abs(it.ts - target) <= measurementToleranceSeconds }.map { it.bpm }
            if (values.size < minimumSamplesPerReading) return null
            return endHr - (median(values) ?: return null)
        }

        return Result(
            endHr = endHr,
            after1Minute = recovery(1),
            after2Minutes = recovery(2),
            after5Minutes = recovery(5),
        ).takeIf { it.hasMeasurement }
    }

    private fun sustainedSeconds(threshold: Double, samples: List<HrSample>): Long {
        if (samples.size < 2) return 0L
        var seconds = 0L
        for (i in 0 until samples.lastIndex) {
            val gap = samples[i + 1].ts - samples[i].ts
            if (gap in 1..maximumContinuousGapSeconds && samples[i].bpm.toDouble() >= threshold) {
                seconds += gap
            }
        }
        return seconds
    }

    private fun median(values: List<Int>): Int? {
        if (values.isEmpty()) return null
        val sorted = values.sorted()
        val middle = sorted.size / 2
        return if (sorted.size % 2 == 0) {
            ((sorted[middle - 1] + sorted[middle]) / 2.0).roundToInt()
        } else {
            sorted[middle]
        }
    }
}
